### PR TITLE
Tasks 91-104

### PR DESCRIPTION
**Pull request by KamilWasilonek**

### CODE REVIEW

File: tasks/100_greatest-common-divisor/greatest-common-divisor.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+export function greatestCommonDivisor(...arg) {
+  arg = arg[0] > arg[1] ? arg.reverse() : arg;
```
Comment (tkachukivan) [inline tasks/100_greatest-common-divisor/greatest-common-divisor.js]: Euclidan Algorithm will looks better here

File: tasks/101_least-common-multiple/least-common-multiple.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+export function leastCommonMultiple(arg1, arg2) {
+  let firstArgMultiply = arg1;
```
Comment (tkachukivan) [inline tasks/101_least-common-multiple/least-common-multiple.js]: there is a formula with using greatestCommonDivisor, with it solution can be better and probably faster

File: tasks/102_linked-list/linked-list.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  } else {
+    let current = this.head;
+    while (current.next) {
+      current = current.next;
+    }
+    current.next = node;
+    node.prev = current;
+    this.tail = node;
+  }
+  this.amount += 1;
```
Comment (tkachukivan) [inline tasks/102_linked-list/linked-list.js]: `this.amount++`

File: tasks/102_linked-list/linked-list.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    current = current.next;
+  }
+  return false;
+};
+
+LinkedList.prototype.removeItem = function(value) {
+  let current = this.head;
+  while (current) {
+    if (current.value === value) {
+      if (current == this.head && current == this.tail) {
```
Comment (tkachukivan) [inline tasks/102_linked-list/linked-list.js]: `===` instead of `==`

File: tasks/103_stack/stack.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  this.last = null;
+  this.size = 0;
+  this.isEmpty = true;
+}
+
+Stack.prototype.push = function(data) {
+  let node = new Node(data);
+  if (this.last === null) {
+    this.last = node;
+    this.size += 1;
```
Comment (tkachukivan) [inline tasks/103_stack/stack.js]: in both cases you update size and isEmpty, maybe you can move this outside of your ifs to not repeat yourself

File: tasks/103_stack/stack.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  }
+};
+
+Stack.prototype.pop = function() {
+  if (this.last === null) {
+    return undefined;
+  } else {
+    let result = this.last.value;
+    this.last = this.last.prev;
+    this.size -= 1;
```
Comment (tkachukivan) [inline tasks/103_stack/stack.js]: this.size--

File: tasks/91_inventory-update/inventory-update.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    if (!result[item[1]]) {
+      result[item[1]] = item[0];
+    } else {
+      result[item[1]] += item[0];
+    }
+    return result;
+  }, {});
+  return Object.keys(joinedArr)
+      .sort((a, b) => {
+        return a.localeCompare(b);
```
Comment (tkachukivan) [inline tasks/91_inventory-update/inventory-update.js]: 👍 

but one point one not in one line `(a, b) => a.localeCompare(b)`

File: tasks/93_make-a-person/make-a-person.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
+export class Person {
+  constructor(fullName) {
+    this.fullName = fullName;
+    fullName = fullName.split(' ');
```
Comment (tkachukivan) [inline tasks/93_make-a-person/make-a-person.js]: other way would be call setFullName function here, and don't keep fullName but keep first and second, it is easier to concatenate two strings that split every time one

File: tasks/95_pairwise/pairwise.js [hunk 9]
```diff
@@ -0,0 +1,12 @@
+export function pairwise(arr, num) {
+  return arr.reduce((sum, item, index) => {
+    for (let i = index + 1; i < arr.length; i++) {
+      if (item + arr[i] === num) {
+        arr[index] = '';
```
Comment (tkachukivan) [inline tasks/95_pairwise/pairwise.js]: what is a reason to reset current value? you will not return to this value anyway

File: tasks/96_how-many-differences-of-squares/how-many-differences-of-squares.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+export function squares(howManyNumbers) {
+  let result = 0;
+  for (let i = 1; i <= howManyNumbers; i++) {
```
Comment (tkachukivan) [inline tasks/96_how-many-differences-of-squares/how-many-differences-of-squares.js]: are you sure that you need loop here

File: tasks/99_fat-fingers/fat-fingers.js [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+        if (/[aA]/.test(letter)) {
+          shouldCapitalize = !shouldCapitalize;
+        }
+        if (shouldCapitalize) {
+          return letter.toUpperCase();
+        }
+        return letter;
+      })
+      .join('');
+  transformedStr = transformedStr.replace(/[aA]/g, '');
```
Comment (tkachukivan) [inline tasks/99_fat-fingers/fat-fingers.js]: what about reduce method and do it without replace?

**Review by tkachukivan — CHANGES_REQUESTED** [review]
please make changes to 96, 99 and 100/101 optional :)

File: tasks/93_make-a-person/make-a-person.js [hunk 12]
```diff
@@ -7,25 +7,23 @@ export class Person {
...
   }
 
   setFirstName(first) {
     this._firstName = first;
-    this.fullName = this._firstName + ' ' + this._lastName;
   }
 
   setLastName(last) {
     this._lastName = last;
-    this.fullName = this._firstName + ' ' + this._lastName;
```
Comment (tkachukivan) [inline tasks/93_make-a-person/make-a-person.js]: with this changes when you set fullName, your first and last name remains as they was

**Review by tkachukivan — CHANGES_REQUESTED** [review]

File: tasks/103_stack/stack.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+Stack.prototype.push = function(data) {
+  let node = new Node(data);
+  if (this.last === null) {
+    this.last = node;
+  } else {
+    node.prev = this.last;
+    this.last = node;
+  }
```
Comment (tsvetkovv) [inline tasks/103_stack/stack.js:16]: repeated code in both of branches. Can be simplified

File: tasks/103_stack/stack.js [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+  this.size++;
+  this.isEmpty = false;
+};
+
+Stack.prototype.pop = function() {
+  if (this.last === null) {
+    return undefined;
+  } else {
```
Comment (tsvetkovv) [inline tasks/103_stack/stack.js]: `else` is not required here - `return` is in the first branch of condition

File: tasks/93_make-a-person/make-a-person.js [hunk 15]
```diff
@@ -0,0 +1,32 @@
+export class Person {
+  constructor(fullName) {
+    this.fullName = fullName;
+    const dividedFullName = fullName.split(' ');
+    this._firstName = dividedFullName[0];
```
Comment (tsvetkovv) [inline tasks/93_make-a-person/make-a-person.js]: you have setter for it

File: tasks/93_make-a-person/make-a-person.js [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  setLastName(last) {
+    this._lastName = last;
+  }
+
+  setFullName(fullName) {
+    this.fullName = fullName;
+  }
+}
```
Comment (tsvetkovv) [inline tasks/93_make-a-person/make-a-person.js:30]: You have dependencies: 

1. first name depends on full name and vice versa. If you set first name, you should update fullName as well and vice versa
2. last name depends on full name and vice versa. If you set last name, you should update fullName as well and vice versa

unfortunately, tests are not covered this case. But it's real production case. It is too hard to keep up to date both source of truth, so please try to **have only one source of truth**. For example, only first + last name in this class

**steps to reproduce**: set  another full name, get first name.
**expected result**: first name should be updated
The same for second name.

Comment (tsvetkovv) [reply tasks/93_make-a-person/make-a-person.js:30]: And partial test does not work
image

File: tasks/95_pairwise/pairwise.js [hunk 17]
```diff
@@ -0,0 +1,12 @@
+export function pairwise(arr, num) {
+  let arrToSearch = arr;
+  return arr.reduce((sum, item, index) => {
+    for (let i = index + 1; i < arr.length; i++) {
+      if (item + arr[i] === num) {
+        arrToSearch[i] = '';
```
Comment (tsvetkovv) [inline tasks/95_pairwise/pairwise.js:6]: here `arrToSearch[i] === arr[i]`
and `arrToSearch === arr`. It's the same object in memory

The method what calls `pairwise` does not expect that input array would be changed

File: tasks/95_pairwise/pairwise.js [hunk 18]
```diff
@@ -0,0 +1,12 @@
+export function pairwise(arr, num) {
+  let arrToSearch = arr;
```
Comment (tsvetkovv) [inline tasks/95_pairwise/pairwise.js]: btw, what this variable for? It has never read

File: tasks/95_pairwise/pairwise.js [hunk 19]
```diff
@@ -0,0 +1,12 @@
+export function pairwise(arr, num) {
```
Comment (tsvetkovv) [inline tasks/95_pairwise/pairwise.js]: num - sum?

Comment (tsvetkovv) [reply tasks/95_pairwise/pairwise.js]: or `expectedSum`

**Review by tsvetkovv — CHANGES_REQUESTED** [review]

**Review by tkachukivan — APPROVED** [review]

**Review by tsvetkovv — APPROVED** [review]
Well done!

**Pull request merged**